storm-client/test/jvm/org/apache/storm/topology/PersistentWindowedBoltExecutorTest.java

```diff
@@ -0,0 +1,261 @@
+        Mockito.when(mockPartitionState.get(Mockito.any(), Mockito.any())).then(returnsArgAt(1));
+        Mockito.when(mockWindowState.get(Mockito.any(), Mockito.any())).then(returnsArgAt(1));
+        Mockito.when(mockSystemState.get(Mockito.any(), Mockito.any())).then(returnsArgAt(1));
+        executor.prepare(mockStormConf, mockTopologyContext, mockOutputCollector,
+            mockWindowState, mockPartitionState, mockSystemState);
+    }
+
+    @Test
+    public void testExecuteTuple() throws Exception {
+        Mockito.when(mockWaterMarkEventGenerator.track(Mockito.any(GlobalStreamId.class), Mockito.anyLong())).thenReturn(true);
+        Tuple mockTuple = Mockito.mock(Tuple.class);
+        executor.initState(null);
+        executor.waterMarkEventGenerator = mockWaterMarkEventGenerator;
+        executor.execute(mockTuple);
+        // should be ack-ed once
+        Mockito.verify(mockOutputCollector, Mockito.times(1)).ack(mockTuple);
+    }
+
+    @Test
+    public void testExecuteLatetuple() throws Exception {
+        Mockito.when(mockWaterMarkEventGenerator.track(Mockito.any(GlobalStreamId.class), Mockito.anyLong())).thenReturn(false);
+        Tuple mockTuple = Mockito.mock(Tuple.class);
+        executor.initState(null);
+        executor.waterMarkEventGenerator = mockWaterMarkEventGenerator;
+        executor.execute(mockTuple);
+        // ack-ed once
+        Mockito.verify(mockOutputCollector, Mockito.times(1)).ack(mockTuple);
+        // late tuple emitted
+        Mockito.verify(mockOutputCollector, Mockito.times(1))
+            .emit(stringCaptor.capture(), anchorCaptor.capture(), valuesCaptor.capture());
+        Assert.assertEquals(LATE_STREAM, stringCaptor.getValue());
+        Assert.assertEquals(Collections.singletonList(mockTuple), anchorCaptor.getValue());
+        Assert.assertEquals(new Values(mockTuple), valuesCaptor.getValue());
+    }
+
+    @Test
+    public void testActivation() throws Exception {
+        Mockito.when(mockWaterMarkEventGenerator.track(Mockito.any(GlobalStreamId.class), Mockito.anyLong())).thenReturn(true);
+        executor.initState(null);
+        executor.waterMarkEventGenerator = mockWaterMarkEventGenerator;
+
+        List<Tuple> mockTuples = getMockTuples(5);
+        mockTuples.forEach(t -> executor.execute(t));
+        // all tuples acked
+        Mockito.verify(mockOutputCollector, Mockito.times(5)).ack(tupleCaptor.capture());
+        Assert.assertArrayEquals(mockTuples.toArray(), tupleCaptor.getAllValues().toArray());
+
+        Mockito.doAnswer(new Answer<Void>() {
+            @Override
+            public Void answer(InvocationOnMock invocation) throws Throwable {
+                TupleWindow window = (TupleWindow) invocation.getArguments()[0];
+                // iterate the tuples
+                Assert.assertEquals(5, window.get().size());
+                // iterating multiple times should produce same events
```

Comment:
Am pulling out the WindowState into a top level class and adding additional tests. Maybe this can still be left as is.